lib/runner.js

```diff
@@ -161,12 +169,31 @@ var prepare = () => () => {
     });
 };
 
+var sourceDirectories = args => {
+    var elmJsonPath = path.join(args.path, 'elm.json');
+    return fs.readJson(elmJsonPath)
+        .then(elmJson => elmJson['source-directories']);
+};
+
 var allSources = args => {
+    return sourceDirectories(args)
+        .then(sourceDirs => Promise.all(sourceDirs.map(sourceDir => allSourcesFromDir(sourceDir))))
+        .then(sourcesArrays => sourcesArrays.flat());
+};
+
+var allSourcesFromDir = dir => {
     return new Promise(function(resolve, reject) {
-        find.file(/\.(elm|js)$/, args.path, resolve);
+        find.file(/\.(elm|js)$/, dir, resolve);
     });
 };
 
+var fixOutOfRootPath = path => {
+  // TODO this is too simplistic (we handle ../x but not foo/../../x nor many other cases)
+  return path.startsWith('..') 
+    ? path.replace(/^\.\./,'__dotdot')
+    : path;
+}
```

Comment:
Our code (that I'm hacking `elm-coverage` for) has this directory structure:
```
├── settings
│   ├── elm.json
│   ├── elm-stuff
│   ├── src
│   └── tests
├── _share
│   ├── benchmarks
│   ├── elm.json
│   ├── elm-stuff
│   ├── src
│   └── tests
├── tv-study
│   ├── elm.json
│   ├── elm-stuff
│   ├── src
│   └── tests
```
and some of these `elm.json`s depend on each other, so we have:
```
    "source-directories": [
        "../_factories",
        "../_share/src",
        "src"
    ],
```
It's because of this case that I'm replacing `..` with `__dotdot`. If we wanted to blindly copy the paths otherwise, it would leak out of the `.coverage/instrumented` folder. This way it stays inside, like `.coverage/instrumented/__dotdot/_factories/Foo.elm`

As the TODO comment says, it doesn't solve this very comprehensively (but it works for my case, so I stopped. Sorry :) Again, I don't think this PR should be accepted, maybe think of it just as a "lessons learned" point for if you ever want to do this properly)